lib/swtpmtest.pm

```diff
@@ -0,0 +1,115 @@
+# Copyright (C) 2020-2021 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+#
+# Summary: Base module for swtpm test cases
+# Maintainer: rfan1 <richard.fan@suse.com>
+
+package swtpmtest;
+
+use base Exporter;
+use Exporter;
+use strict;
+use warnings;
+use testapi;
+use utils;
+
+our @EXPORT = qw(
+  start_swtpm_vm
+  stop_swtpm_vm
+  swtpm_verify
+);
+
+my $image_path  = '/var/lib/libvirt/images';
+my $guestvm_xml = {
+    swtpm_1 => {
+        xml_file    => {uefi => 'swtpm_uefi_1_2.xml', legacy => 'swtpm_legacy_1_2.xml'},
+        version     => '1.2',
+        ssh_script  => 'ssh_script_1',
+        result_file => '/tmp/swtpm_1',
+    },
+    swtpm_2 => {
+        xml_file    => {uefi => 'swtpm_uefi_2_0.xml', legacy => 'swtpm_legacy_2_0.xml'},
+        version     => '2.0',
+        ssh_script  => 'ssh_script_2',
+        result_file => '/tmp/swtpm_2',
+    },
+};
+
+my $sample_cfg = {
+    uefi   => {vm_name => 'SLES15SP3-UEFI-swtpm',   sample_file => 'swtpm_uefi.xml'},
+    legacy => {vm_name => 'SLES15SP3-LEGACY-swtpm', sample_file => 'swtpm_legacy.xml'},
+};
+
+sub start_swtpm_vm {
+    my ($swtpm_ver, $swtpm_vm_type) = @_;
+    die "invalid swtpm parameter"   unless $guestvm_xml->{$swtpm_ver};
+    die "invalid vm mode parameter" unless $sample_cfg->{$swtpm_vm_type};
+
+    # Copy the sample configuration file and modify it based on the test requirement
+    my $operate_item  = $guestvm_xml->{$swtpm_ver};
+    my $operate_item1 = $sample_cfg->{$swtpm_vm_type};
+    assert_script_run("cd $image_path");
+    assert_script_run("cp $operate_item1->{sample_file} $operate_item->{xml_file}->{$swtpm_vm_type}");
+    assert_script_run(
+"sed -i \"/<\\/devices>/i\\    <tpm model='tpm-tis'>\\n      <backend type='emulator' version='$operate_item->{version}'\\/>\\n    <\\/tpm>\" $operate_item->{xml_file}->{$swtpm_vm_type}"
+    );
+    assert_script_run("virsh define $operate_item->{xml_file}->{$swtpm_vm_type}");
+    assert_script_run("virsh start $operate_item1->{vm_name}");
+}
+
+sub stop_swtpm_vm {
+    my $para = shift;
+    die "invalid vm type parameter" unless $sample_cfg->{$para};
+    my $operate_item = $sample_cfg->{$para};
```

Comment:
is it should be `$operate_item1` ?  Perhaps it should be worked here. Just curious, I am just wondering why if it is not the corresponding name within start_swtpm_vm()

Reply:
let me modify it

Reply:
done